sdk/python/starwhale/data_store.py

```diff
@@ -0,0 +1,703 @@
+FLOAT32 = Type("float", pa.float32(), 32, 0.0)
+FLOAT64 = Type("float", pa.float64(), 64, 0.0)
+BOOL = Type("bool", pa.bool_(), 1, 0)
+STRING = Type("str", pa.string(), 32, "")
+BYTES = Type("bytes", pa.binary(), 32, b"")
+LINK = LinkType()
+
+_TYPE_DICT: dict[Type, Type] = {
+    type(None): NONE,
+    np.byte: INT8,
+    np.int8: INT8,
+    np.int16: INT16,
+    np.int32: INT32,
+    np.int64: INT64,
+    int: INT64,
+    np.float16: FLOAT16,
+    np.float32: FLOAT32,
+    np.float_: FLOAT64,
+    float: FLOAT64,
+    np.bool_: BOOL,
+    bool: BOOL,
+    str: STRING,
+    bytes: BYTES,
+    Link: LINK,
+}
+
+_TYPE_NAME_DICT = {str(v): v for k, v in _TYPE_DICT.items()}
+
+
+def _get_type(obj: Any) -> Type:
+    return _TYPE_DICT.get(type(obj), None)
+
+
+class ColumnSchema:
+    def __init__(self, name: str, type: Type):
+        self.name = name
+        self.type = type
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, ColumnSchema) and self.name == other.name and self.type is other.type
+
+
+class TableSchema:
+    def __init__(self, key_column: str, columns: list[ColumnSchema]):
+        self.key_column = key_column
+        self.columns = {col.name: col for col in columns}
+
+    def copy(self) -> "TableSchema":
+        return TableSchema(self.key_column, self.columns.values())
+
+    @staticmethod
+    def parse(json_str: str) -> "TableSchema":
+        d = json.loads(json_str)
+        return TableSchema(d["key"], [ColumnSchema(col["name"], _TYPE_NAME_DICT[col["type"]]) for col in d["columns"]])
+
+    def __str__(self) -> str:
+        return json.dumps(
+            {
+                "key": self.key_column,
+                "columns": [{"name": col.name, "type": str(col.type)} for col in self.columns.values()],
+            }
+        )
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, TableSchema) and self.key_column == other.key_column and self.columns == other.columns
+
+
+def _get_table_path(root_path: str, table_name: str) -> str:
+    return os.path.join(root_path, *table_name.split("/"))
+
+
+def _parse_parquet_name(name: str) -> Tuple[str, int]:
+    try:
+        if name.endswith(".parquet"):
+            if name.startswith("base-"):
+                return "base", int(name[5:-8])
+            elif name.startswith("patch-"):
+                return "patch", int(name[6:-8])
+    except ValueError:
+        # ignore invalid filename
+        pass
+    return None, None
+
+
+def _write_parquet_file(filename: str, table: pa.Table) -> None:
+    with pq.ParquetWriter(filename, table.schema) as writer:
+        writer.write_table(table)
+
+
+def _scan_parquet_file(
+    path: str, columns: Optional[dict[str, str]] = None, start: Optional[Any] = None, end: Optional[Any] = None
+) -> Iterator[dict]:
+    f = pq.ParquetFile(path)
+    schema_arrow = f.schema_arrow
+    if columns is None:
```

Comment:
does the empty colums dict should be set the `{name: name for name in schema_arrow.names if name != "-" and not name.startswith("~")}` ?

```
if columns is None:
    columns = {name: name for name in schema_arrow.names if name != "-" and not name.startswith("~")}
```
-->
```
columns = columns or {name: name for name in schema_arrow.names if name != "-" and not name.startswith("~")}
```

Reply:
columns should not be empty here. As it is an optional parameter, users should provide a meaningful value when they specify it explicitly. So I would like to return an empty list for empty columns rather than treat it as the same as None.